base_layer/core/src/consensus/consensus_encoding/hash_writer.rs

```diff
@@ -23,26 +23,34 @@
 use std::io::Write;
 
 use digest::{consts::U32, Digest, FixedOutput, Update};
-use tari_common_types::types::HashDigest;
+use tari_crypto::{hash::blake2::Blake256, hashing::AsFixedBytes};
 
-use crate::consensus::ConsensusEncoding;
+use crate::consensus::{ConsensusEncoding, ConsensusEncodingDomainHasher};
 
-#[derive(Clone)]
 pub struct ConsensusHashWriter<H> {
-    digest: H,
+    hasher: ConsensusEncodingDomainHasher<H>,
+    internal_state: Vec<u8>,
```

Comment:
One of the primary goals of the consensus hash writer is to have zero heap allocations since hashes are done very often. This will result in many allocations as each element is added. I understand why you did this (due to the length being added to the domain-separated hash for each update). I suggest we simply add a static domain separation label to the consensus hashing and use a regular hash digest since consensus encoding already has canonical encoding/length encoding where applicable.

Reply:
Seems like a good idea